viewing_party/party.py

```diff
@@ -1,23 +1,196 @@
+
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+# user_data = {watchlist: ; watched: }
+# title = "movie user has watched"
+# if title is in watchlist, remove from watchlist, add to watched, return user_data.
+# if title not in watchlist, return user_data.
+
+    for movie in user_data["watchlist"]:
+        if title == movie["title"]:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+            
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    total_rating = 0.0
+
+    if not user_data["watched"]:
+        return total_rating
+    
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"]
+    
+    average_rating = total_rating / len(user_data["watched"])
+    
+    return average_rating
+
+
+def get_most_watched_genre(user_data):
+    
+    if user_data["watched"] == []:
+        return None
+    
+    genre_dict = {}
+
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in genre_dict:
+            genre_dict[genre] += 1
+        else:
+            genre_dict[genre] = 1
+        
+    highest_counter = 0
+
+    for genre in genre_dict:
+        if genre_dict[genre] > highest_counter:
+            highest_counter = genre_dict[genre]
+            most_frequent_genre = genre
+
+    return most_frequent_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+# user_data = {"watched": [{}], "friends": [{"watched":[{}]}]}
+# return movie user has watched but friends haven't as a list of dictionaries
+# compare {} in "watched" and "friends":"watched", return the unique one in "watched".
+
+# make a list of friends wathced movie titles. Set makes it more efficient.
+    friends_movies = set()
+    for movie_dict in user_data["friends"]:
+        for movie in movie_dict["watched"]:
+            friends_movies.add(movie["title"])
+
+# loop throught user watched list, return ones different than friends watched list.
+    user_unique = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
+
+    return user_unique       
+
+def get_friends_unique_watched(user_data):
+# return movie at least one of the friends have watched, but user hasn't as a list of dict
+    user_movies = set()
+    for movie in user_data["watched"]:
+        user_movies.add(movie["title"])
 
+    friends_unique = []
+    for movie_dict in user_data["friends"]:
+        for movie in movie_dict["watched"]:
+            if movie["title"] not in user_movies and movie not in friends_unique:
+                friends_unique.append(movie)

```

Comment:
This is a bit of a tough set of nested loops and conditionals to avoid, unfortunately given the way our data is structured! Great job working with what y'all have!